src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -170,6 +173,13 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
             // doc events and appropriately map files to/from it and other relevant workspaces (like the
             // metadata-as-source workspace).
             await this.ComponentModel.GetService<MiscellaneousFilesWorkspace>().InitializeAsync(this).ConfigureAwait(false);
+
+            // Proffer in-process service broker services
+            var serviceBrokerContainer = await this.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>(this.JoinableTaskFactory).ConfigureAwait(false);
+
+            serviceBrokerContainer.Proffer(
+                WorkspaceProjectFactory.ServiceDescriptor,
+                (_, _, _, _) => ValueTaskFactory.FromResult<object?>(new WorkspaceProjectFactory(this.ComponentModel.GetService<IWorkspaceProjectContextFactory>())));
```

Comment:
slightly concerned about costs here.

Reply:
Costs of what?

Reply:
doing this work in package initialize :)

Reply:
I don't think there's an alternative approach, but I generally agree I don't like package initialize in general for things like this. @AArnott is there an alternative?

Reply:
If there's one thing Package.Initialize has always been for, it's proffering its services. That's what this is. And the cost is minimal. The Proffer call is essentially free. The query for the global brokered service container service itself has some (small) startup cost, but it's already loaded by the time you ask for it, almost certainly.
When you register your brokered service, the brokered service container knows the package that proffers it, and it will load the package when an incoming brokered service call requires it. It relies on Package.Init proffering before reporting the package has initialized. If that doesn't happen, the client will get null back instead of your brokered service.

Now, as an aside, I see your service factory just defers to MEF to do the work. Your package's requiring of MEF to initialize is already 1000X more expensive (in worst case, anyway) than the global brokered service container would ever be. So I don't think this is going to be your perf problem.
Also: given your brokered service is a MEF part, did you know that you can drop the proffering code and [just use MEF directly](https://learn.microsoft.com/en-us/visualstudio/extensibility/how-to-provide-brokered-service?view=vs-2022#MEF)?

Reply:
> Also: given your brokered service is a MEF part, did you know that you can drop the proffering code and [just use MEF directly](https://learn.microsoft.com/en-us/visualstudio/extensibility/how-to-provide-brokered-service?view=vs-2022#MEF)?

@AArnott Ah, perfect, I was hoping there was a way like that. (And as an aside, we really should update the internal wiki pages to point to that public docs...I didn't know there were public docs.)

Reply:
Done. I've added a link. Good idea.

If you use MEF to export your brokered service, you should know that during startup, before MEF has pushed registration of brokered services into the global container, a brokered service client may get `null` back. Since your packages initialize MEF as part of package init, it's likely that a null result for you would be rare. But it is a different guarantee than when you register the brokered service yourself via pkgdef. You can do it both: use pkgdef and MEF, in which case the package guid you register with is MEF's instead of your own. Then you don't risk the null near startup.

Reply:
> If you use MEF to export your brokered service, you should know that during startup, before MEF has pushed registration of brokered services into the global container, a brokered service client may get null back.

@AArnott I'm a bit confused here -- I'd expect any exports to simply be pulled in always and there wouldn't be any pkgdef discovery at all. Why isn't it a pull model instead of a push model?

Reply:
A pull model would mean that no brokered services could be produced without first waiting for MEF to initialize to register its exports with the container. But the container is really fast to start right now, and MEF can be quite slow, so we didn't want to make the container slow just for the sake of MEF, particularly since we had many pkgdef-registered brokered services and few or (at the time) 0 MEF brokered services.
Since this bothers you, consider following [this backlog item](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1465661).

Reply:
> Since this bothers you, consider following [this backlog item](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1465661).

@AArnott  I think we're fine for now just pushing it into the container the way we are in this PR; we're just generally wary of anything in package load as it becomes a single point of slowdown during startup for us. Practically any slowdown in that service broker registration becomes a tax on the existing codepaths loading the package, but worse it also means we have to pay various loading costs for other legacy things in our package load just to proffer the service too.

> Since this bothers you

Practically, the fact we trigger MEF today during our package load is something we want to clean up (see prior comment about package loading being an icky source of performance issues), so it seems any effort trying to MEF exported thing is walking on thin ice and would break at some point.

> You can do it both: use pkgdef and MEF, in which case the package guid you register with is MEF's instead of your own. 

Is this officially supported?

Reply:
> That's probably ok, but as you say you're trying to remove MEF from your startup path,

I should say it's not that MEF specifically is a concern in our startup path, as much as historically we've used package initialize as a place to put "things that need to run at start"; the problem being that if you have an initialize method initializing 10 unrelated things, the time any one of those need to run you're also paying a startup cost.

> But seeing as the editor requires MEF, avoiding MEF in a language service may not be ... useful.

It's not just our editor interactions: in Roslyn we discover all sorts of language-specific services via MEF as well. In this case, the service we're implementing here uses MEF too, so that's pretty much assumed.